src/crypto/ecdh.rs

```diff
@@ -145,18 +145,41 @@ pub fn decrypt(priv_key: &ECDHSecretKey, mpis: &[Mpi], fingerprint: &[u8]) -> Re
         .copy_from_slice(encrypted_session_key);
 
     let mut decrypted_key_padded = aes_kw::unwrap(&z, &encrypted_session_key_vec)?;
-    // PKCS5 unpadding (PKCS5 is PKCS7 with a blocksize of 8)
+    // PKCS5-style unpadding (PKCS5 is PKCS7 with a blocksize of 8).
+    //
+    // RFC 6637 describes the padding:
+    // a) "The result is padded using the method described in [PKCS5] to the 8-byte granularity."
+    // b) "For example, assuming that an AES algorithm is used for the session key, the sender MAY
+    // use 21, 13, and 5 bytes of padding for AES-128, AES-192, and AES-256, respectively, to
+    // provide the same number of octets, 40 total, as an input to the key wrapping method."
+    //
+    // So while the padding ensures that the length of the padded message is a multiple of 8, the
+    // padding may exceed 8 bytes in size.
     {
         let len = decrypted_key_padded.len();
         let block_size = 8;
         ensure!(len % block_size == 0, "invalid key length {}", len);
         ensure!(!decrypted_key_padded.is_empty(), "empty key is not valid");
 
-        // grab the last block
-        let offset = len - block_size;
-        let last_block = GenericArray::<u8, U8>::from_slice(&decrypted_key_padded[offset..]);
-        let unpadded_last_block = Pkcs7::unpad(last_block)?;
-        let unpadded_len = offset + unpadded_last_block.len();
+        // The last byte should contain the padding symbol, which is also the padding length
+        let pad = decrypted_key_padded.last().expect("is not empty");
+
+        // Padding length seems to exceed size of the padded message
+        if *pad as usize > len {
```

Comment:
I believe the padding scheme in RFC 6637 explicitly allows the padding to exceed block length. The rationale is to obfuscate the length of the symmetric key.

For example, RFC 6637 mentions 21 bytes of padding as a legal format for an AES128 key (also see my comment in https://github.com/rpgp/rpgp/pull/280/files#diff-962d2eaeb82df8d56721a98d7d86e270a5021cfa2b361b25146ecbd68d47cf51R150).

The unpadded key must have the expected length for the algorithm, of course. The previous code didn't check for this property. I'm not sure if it would be useful to check for that.

My feeling is: this unpadding function should not fail with panics, but if the sender sends us a padded key that is just 1 byte long, then that's what we have. Trying to us that key should fail later in the code, then?

Reply:
Putting a finer point on it: The purpose of this PR is to handle padding that exceeds block length.

Because that currently fails, which causes interop problems when using AES128 with GopenPGP, as well as with bouncy castle-based implementations (like PGPainless).

Reply:
> The unpadded key must have the expected length for the algorithm, of course. The previous code didn't check for this property. I'm not sure if it would be useful to check for that. 

we really should

Reply:
re reading the RFC

>  This encoding allows
   the sender to obfuscate the size of the symmetric encryption key used
   to encrypt the data.  For example, assuming that an AES algorithm is
   used for the session key, the sender MAY use 21, 13, and 5 bytes of
   padding for AES-128, AES-192, and AES-256, respectively, to provide
   the same number of octets, 40 total, as an input to the key wrapping
   method. 

it seems that the padding is not variable but deterministic, depending on the key size, so that you always end up with 40 bytes

Reply:
> it seems that the padding is not variable but deterministic, depending on the key size, so that you always end up with 40 bytes (if the obfuscation is used)

I don't think that's what happens in practice. As far as I recall, not all implementation produce padded AES keys at 40 bytes length.

Note that it's just a "MAY", not a "MUST". Implementations do different padding, but they are allowed to pad *up to* the maximum length.

Reply:
sigh, I hate variable padding…